File: ios/SettingsUtils.swift
//
//  SettingsUtils.swift
//  Pods
//
//  Created by Rakesh Ayyaswami on 23/9/2024.
//

import BRLMPrinterKit


internal class SettingsUtils {
    
    private static func _printerModelFromName(_ name: String) throws -> BRLMPrinterModel {
        switch name {
        case "QL-710W":    return .QL_710W
        case "QL-720NW":   return .QL_720NW
        case "QL-810W":    return .QL_810W
        case "QL-820NWB":  return .QL_820NWB
        case "QL-1110NWB": return .QL_1110NWB
        case "QL-1115NWB": return .QL_1115NWB

        default:
            throw GenericError(description: "Unsupported printer model")
        }
    }
    
    private static func _parseSettings_QLSeries(_ reader: DictionaryReadWrite, model: BRLMPrinterModel) throws -> BRLMQLPrintSettings {
        guard let settings = BRLMQLPrintSettings(defaultPrintSettingsWith: model) else {
            throw GenericError(description: "Failed to initialize printer settings object")
        }
        
        settings.labelSize               = reader.read("labelSize",               nvl: settings.labelSize) { BRLMQLPrintSettingsLabelSize(rawValue: $0) }
        settings.autoCutForEachPageCount = reader.read("autoCutForEachPageCount", nvl: 1)
        settings.autoCut                 = reader.read("autoCut",                 nvl: false)
        settings.cutAtEnd                = reader.read("cutAtEnd",                nvl: false)
        settings.resolution              = reader.read("resolution",              nvl: .normal)            { BRLMPrintSettingsResolution(rawValue: $0) }
        settings.halftone                = reader.read("halftone",                nvl: .threshold)         { BRLMPrintSettingsHalftone(rawValue: $0) }
        settings.halftoneThreshold       = reader.read("halftoneThreshold",       nvl: 128)
        
        return settings
    }
    
    internal static func settingsFromDictionary(_ settings: [String: Any], modelName: String) throws  -> BRLMPrintSettingsProtocol {

        // decode supported model from name
        let model  = try _printerModelFromName(modelName)

        // parse settings based on model
        let reader = DictionaryReadWrite(settings)
        if  modelName.hasPrefix("QL") {
            return try _parseSettings_QLSeries(reader, model: model)
        }
        
        // unknown model
        throw GenericError(description: "Unsupported printer model")
    }
}

File: android/src/main/java/com/vpass/brotherprintersdk/SettingsUtils.kt
package com.vpass.brotherprintersdk

import android.util.Log
import com.brother.sdk.lmprinter.PrinterModel
import com.brother.sdk.lmprinter.setting.PrintImageSettings
import com.brother.sdk.lmprinter.setting.PrintSettings
import com.brother.sdk.lmprinter.setting.QLPrintSettings

class SettingsUtil private constructor() {
  companion object {
    private fun _printerModelFromName(name: String): PrinterModel? {
      if (name.startsWith("QL-710W")) {
        Log.d("ExpoBrotherPrinterSdk", "_printerModelFromName - Found QL-710W")
        return PrinterModel.QL_710W
      }
      if (name.startsWith("QL-720NW")) {
        Log.d("ExpoBrotherPrinterSdk", "_printerModelFromName - Found QL-720NW")
        return PrinterModel.QL_720NW
      }
      if (name.startsWith("QL-810W")) {
        Log.d("ExpoBrotherPrinterSdk", "_printerModelFromName - Found QL-810W")
        return PrinterModel.QL_810W
      }
      if (name.startsWith("QL-820NWB")) {
        Log.d("ExpoBrotherPrinterSdk", "_printerModelFromName - Found QL-820NWB")
        return PrinterModel.QL_820NWB
      }
      if (name.startsWith("QL-1110NWB")) {
        Log.d("ExpoBrotherPrinterSdk", "_printerModelFromName - Found QL-1110NWB")
        return PrinterModel.QL_1110NWB
      }
      if (name.startsWith("QL-1115NWB")) {
        Log.d("ExpoBrotherPrinterSdk", "_printerModelFromName - Found QL-1115NWB")
        return PrinterModel.QL_1115NWB
      }
      Log.d("ExpoBrotherPrinterSdk", "_printerModelFromName - No matching model found for: $name")
      return null;
    }

    private fun _labelSizeFromValue(value: Int): QLPrintSettings.LabelSize {
      return when (value) {
        0 -> QLPrintSettings.LabelSize.DieCutW17H54
        1 -> QLPrintSettings.LabelSize.DieCutW17H87
        2 -> QLPrintSettings.LabelSize.DieCutW23H23
        3 -> QLPrintSettings.LabelSize.DieCutW29H42
        4 -> QLPrintSettings.LabelSize.DieCutW29H90
        5 -> QLPrintSettings.LabelSize.DieCutW38H90
        6 -> QLPrintSettings.LabelSize.DieCutW39H48
        7 -> QLPrintSettings.LabelSize.DieCutW52H29
        8 -> QLPrintSettings.LabelSize.DieCutW62H29
        9 -> QLPrintSettings.LabelSize.DieCutW62H60
        10 -> QLPrintSettings.LabelSize.DieCutW62H75
        11 -> QLPrintSettings.LabelSize.DieCutW62H100
        12 -> QLPrintSettings.LabelSize.DieCutW60H86
        13 -> QLPrintSettings.LabelSize.DieCutW54H29
        14 -> QLPrintSettings.LabelSize.DieCutW102H51
        15 -> QLPrintSettings.LabelSize.DieCutW102H152
        16 -> QLPrintSettings.LabelSize.DieCutW103H164
        17 -> QLPrintSettings.LabelSize.RollW12
        18 -> QLPrintSettings.LabelSize.RollW29
        19 -> QLPrintSettings.LabelSize.RollW38
        20 -> QLPrintSettings.LabelSize.RollW50
        21 -> QLPrintSettings.LabelSize.RollW54
        22 -> QLPrintSettings.LabelSize.RollW62
        23 -> QLPrintSettings.LabelSize.RollW62RB
        24 -> QLPrintSettings.LabelSize.RollW102
        25 -> QLPrintSettings.LabelSize.RollW103
        26 -> QLPrintSettings.LabelSize.DTRollW90
        27 -> QLPrintSettings.LabelSize.DTRollW102
        28 -> QLPrintSettings.LabelSize.DTRollW102H51
        29 -> QLPrintSettings.LabelSize.DTRollW102H152
        30 -> QLPrintSettings.LabelSize.RoundW12DIA
        31 -> QLPrintSettings.LabelSize.RoundW24DIA
        32 -> QLPrintSettings.LabelSize.RoundW58DIA
        else -> QLPrintSettings.LabelSize.DieCutW62H29 // Default value
      }
    }

    private fun _resolutionFromValue(value: Int): PrintImageSettings.Resolution {
      return when (value) {
        0 -> PrintImageSettings.Resolution.Low
        1 -> PrintImageSettings.Resolution.Normal
        2 -> PrintImageSettings.Resolution.High
        else -> PrintImageSettings.Resolution.Normal // Default value
      }
    }

    private fun _halftoneFromValue(value: Int): PrintImageSettings.Halftone {
      return when (value) {
        0 -> PrintImageSettings.Halftone.Threshold
        1 -> PrintImageSettings.Halftone.ErrorDiffusion
        2 -> PrintImageSettings.Halftone.PatternDither
        else -> PrintImageSettings.Halftone.Threshold // Default value
      }
    }

    fun settingsFromDictionary(dictionary: Map<String, Any>, modelName: String): PrintSettings {
      // decode supported model from name
      var model = _printerModelFromName(modelName)
      if (model == null) {
        throw GenericError("Unsupported printer model: $modelName")
      }

      // construct default settings for the model
      val settings = QLPrintSettings(model)

      // parse settings from dictionary
      dictionary.forEach { (key, value) ->
        when (key) {
          "labelSize" -> {
            if (value is Int) {
              settings.labelSize = _labelSizeFromValue(value)
            }
          }
          "autoCutForEachPageCount" -> {
            settings.autoCutForEachPageCount = if (value is Int) value else 1
          }
          "autoCut" -> {
            settings.isAutoCut = if (value is Boolean) value else false
          }
          "cutAtEnd" -> {
            settings.isCutAtEnd = if (value is Boolean) value else false
          }
          "resolution" -> {
            if (value is Int) {
              settings.resolution = _resolutionFromValue(value)
            }
          }
          "halftone" -> {
            if (value is Int) {
              settings.halftone = _halftoneFromValue(value)
            }
          }
          "halftoneThreshold" -> {
            if (value is Int) {
              settings.halftoneThreshold = value
            }
          }
        }
      }

      return settings
    }
  }
}
